cirq-core/cirq/sim/classical_simulator.py

```diff
@@ -12,96 +12,134 @@
-from cirq.circuits.circuit import AbstractCircuit
+
+from typing import Dict, Generic, Any, Sequence, TYPE_CHECKING, List
+from cirq import ops, qis
+from cirq.value import big_endian_int_to_bits
 from cirq.ops.raw_types import Qid
+from cirq import sim
+from cirq.sim.simulation_state import TSimulationState, SimulationState
 import numpy as np
 
+if TYPE_CHECKING:
+    import cirq
+
+
+class ClassicalBasisState(qis.QuantumStateRepresentation):
+    def __init__(self, initial_state: List[int]):
+        self.basis = initial_state
+
+    def copy(self, deep_copy_buffers: bool = True) -> 'ClassicalBasisState':
+        basis_copy = self.basis if deep_copy_buffers else self.basis.copy()
+        return ClassicalBasisState(basis_copy)
+
+    def measure(self, axes: Sequence[int], seed: 'cirq.RANDOM_STATE_OR_SEED_LIKE' = None):
+        return [self.basis[i] for i in axes]
+
+
+class ClassicalBasisSimState(SimulationState[ClassicalBasisState]):
+    def __init__(self, initial_state, qubits, classical_data):
+        state = ClassicalBasisState(big_endian_int_to_bits(initial_state, bit_count=len(qubits)))
+        super().__init__(state=state, qubits=qubits, classical_data=classical_data)
+
+    def _act_on_fallback_(self, action, qubits: Sequence[Qid], allow_decompose: bool = True):
+        gate = action.gate if isinstance(action, ops.Operation) else action
+        mapped_qubits = [self.qubit_map[i] for i in qubits]
+        if self._is_identity(gate):
+            return True
+        if gate == ops.X:
+            (q,) = mapped_qubits
+            self._state.basis[q] ^= 1
+            return True
+        elif gate == ops.CNOT:
+            c, q = mapped_qubits
+            self._state.basis[q] ^= self._state.basis[c]
+            return True
+        elif gate == ops.SWAP:
+            a, b = mapped_qubits
+            self._state.basis[a], self._state.basis[b] = self._state.basis[b], self._state.basis[a]
+            return True
+        elif gate == ops.TOFFOLI:
+            c1, c2, q = mapped_qubits
+            self._state.basis[q] ^= self._state.basis[c1] & self._state.basis[c2]
+            return True
+        else:
+            raise ValueError(f'{gate} is not one of X, CNOT, SWAP, CCNOT, or a measurement')
+
+    def _is_identity(self, action) -> bool:
+        gate = action.gate if isinstance(action, ops.Operation) else action
+        if isinstance(gate, (ops.XPowGate, ops.CXPowGate, ops.CCXPowGate, ops.SwapPowGate)):
+            return gate.exponent % 2 == 0
+        return False
+
+
+class ClassicalStateStepResult(
+    sim.StepResultBase['ClassicalBasisSimState'], Generic[TSimulationState]
+):
+    """The step result provided by `ClassicalStateSimulator.simulate_moment_steps`."""
+
+
+class ClassicalStateTrialResult(
+    sim.SimulationTrialResultBase['ClassicalBasisSimState'], Generic[TSimulationState]
+):
+    """The trial result provided by `ClassicalStateSimulator.simulate`."""
+
+    def __init__(
```

Comment:
doesn't need an explicit constructor since this class is just a facade of its parent class